## Day 81 Lecture 2 Assignment

In this assignment, we will learn about Dask dataframes. We will use the census dataset loaded below and perform some exploratory data analysis.

In [79]:
!pip install dask[complete] --quiet 
!pip install aiohttp --quiet

In [80]:
import dask.dataframe as dd
import numpy as np

In [81]:
census = dd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/Data%20Sets%20Big%20Data/acs2015_census_tract_data.csv')

In [82]:
census.head()

CensusTract    State   County  ...  SelfEmployed  FamilyWork  Unemployment
0   1001020100  Alabama  Autauga  ...           4.6         0.0           5.4
1   1001020200  Alabama  Autauga  ...           6.1         0.0          13.3
2   1001020300  Alabama  Autauga  ...          12.3         0.0           6.2
3   1001020400  Alabama  Autauga  ...           3.1         0.0          10.8
4   1001020500  Alabama  Autauga  ...           5.3         0.0           4.2

[5 rows x 37 columns]

The dataset described above contains the aggregated data for each county in the United States. For example, the number in the Carpool column describes the percent of people in the county who carpool to work.

Print all the column names in the dataset and describe the data in each column

In [83]:
census.columns

Index(['CensusTract', 'State', 'County', 'TotalPop', 'Men', 'Women',
       'Hispanic', 'White', 'Black', 'Native', 'Asian', 'Pacific', 'Citizen',
       'Income', 'IncomeErr', 'IncomePerCap', 'IncomePerCapErr', 'Poverty',
       'ChildPoverty', 'Professional', 'Service', 'Office', 'Construction',
       'Production', 'Drive', 'Carpool', 'Transit', 'Walk', 'OtherTransp',
       'WorkAtHome', 'MeanCommute', 'Employed', 'PrivateWork', 'PublicWork',
       'SelfEmployed', 'FamilyWork', 'Unemployment'],
      dtype='object')

In [84]:
census.describe().compute()

CensusTract      TotalPop  ...    FamilyWork  Unemployment
count  7.400100e+04  74001.000000  ...  73194.000000  73199.000000
mean   2.839113e+10   4325.591465  ...      0.169772      9.028663
std    1.647593e+10   2129.306903  ...      0.458227      5.955441
min    1.001020e+09      0.000000  ...      0.000000      0.000000
25%    1.303901e+10   2891.000000  ...      0.000000      5.100000
50%    2.804700e+10   4063.000000  ...      0.000000      7.700000
75%    4.200341e+10   5442.000000  ...      0.000000     11.400000
max    7.215375e+10  53812.000000  ...     26.500000    100.000000

[8 rows x 35 columns]

How many counties have a unemployement rate above 10%?

In [85]:
np.count_nonzero(census.groupby("County").mean()["Unemployment"].compute().values>10)

533

There are 533 counties that have an unemployment rate above 10%.

Print the number of counties in each state.

In [87]:
census.groupby("State")["County"].nunique().compute()

State
Alabama                  67
Alaska                   29
Arizona                  15
Arkansas                 75
California               58
Colorado                 64
Connecticut               8
Delaware                  3
District of Columbia      1
Florida                  67
Georgia                 159
Hawaii                    5
Idaho                    44
Illinois                102
Indiana                  92
Iowa                     99
Kansas                  105
Kentucky                120
Louisiana                64
Maine                    16
Maryland                 24
Massachusetts            14
Michigan                 83
Minnesota                87
Mississippi              82
Missouri                115
Montana                  56
Nebraska                 93
Nevada                   17
New Hampshire            10
New Jersey               21
New Mexico               33
New York                 62
North Carolina          100
North Dakota             53
Ohio          

Add a high child poverty boolean column. This column will be 1 if the rate of child poverty is over %20 and 0 otherwise.

In [90]:
census['HighChildPoverty'] = census['ChildPoverty'].map(lambda x: 1 if x>20 else 0)

census.compute().head(10)

CensusTract    State   County  ...  FamilyWork  Unemployment  HighChildPoverty
0   1001020100  Alabama  Autauga  ...         0.0           5.4                 0
1   1001020200  Alabama  Autauga  ...         0.0          13.3                 1
2   1001020300  Alabama  Autauga  ...         0.0           6.2                 0
3   1001020400  Alabama  Autauga  ...         0.0          10.8                 0
4   1001020500  Alabama  Autauga  ...         0.0           4.2                 0
5   1001020600  Alabama  Autauga  ...         0.0          10.9                 1
6   1001020700  Alabama  Autauga  ...         0.0          11.4                 1
7   1001020801  Alabama  Autauga  ...         0.0           8.2                 1
8   1001020802  Alabama  Autauga  ...         0.0           8.7                 0
9   1001020900  Alabama  Autauga  ...         0.0           7.2                 0

[10 rows x 38 columns]

What are the top 10 most populated counties?

In [91]:
census.groupby("County").sum()["TotalPop"].compute().sort_values(ascending=False).head(10)

County
Los Angeles    10038388
Cook            5258620
Orange          5025603
Harris          4389138
Maricopa        4018143
Montgomery      3635047
Jefferson       3623682
San Diego       3223096
Clark           2852761
Kings           2746257
Name: TotalPop, dtype: int64

What is the current number of partitions? 

In [92]:
census.npartitions

1

As it stands, since I'm still using the default init params, I only have 1 partition. It is as if I'm still using Pandas DataFrame.